### PR TITLE
[opt](nereids)push down filter through window

### PR DESCRIPTION
**Pull request by englefly**
## Proposed changes
select rank() over (partition by A, B) as r, sum(x) over(A, C) as s from T;
      A is a common partition key for all windowExpressions, that is A is intersection of {A,B} and {A, C}
we could push filter A=1 through this window, since  A is a common Partition key:
`select * from (select a, row_number() over (partition by a) from win) T where a=1;`
origin plan:
```
----filter((T.a = 1))
----------PhysicalWindow
------------PhysicalQuickSort
--------------PhysicalProject
------------------PhysicalOlapScan[win]
```

transformed to 
```
----PhysicalWindow
------PhysicalQuickSort
--------PhysicalProject
----------filter((T.a = 1))
------------PhysicalOlapScan[win]
```

But C=1 can not be pushed through window.

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalWindow.java [hunk 1]
```diff
@@ -224,4 +229,31 @@ public Optional<Plan> pushPartitionLimitThroughWindow(long partitionLimit, boole
...
     }
+
+    /**
+     *
+     * select rank() over (partition by A, B) as r, sum(x) over(A, C) as s from T;
+     * A is a common partition key for all windowExpressions.
+     * for a common Partition key A, we could push filter A=1 through this window.
+     */
+    public Set<Expression> getCommonPartitionKeyFromWindowExpressions() {
+        Set<Expression> commonPartitionKeySet = Sets.newHashSet();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalWindow.java:240]: use ImmutableSet.Builder to construct a set

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalWindow.java [hunk 2]
```diff
@@ -224,4 +229,31 @@ public Optional<Plan> pushPartitionLimitThroughWindow(long partitionLimit, boole
...
+        Map<Expression, Integer> partitionKeyCount = Maps.newConcurrentMap();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalWindow.java:241]: why concurrentMap?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushdownFilterThroughWindow.java
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushdownFilterThroughWindow.java:1]: i think we should not only use regression test but also use fe ut to cover more scenario and ensure all code branch is right

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushdownFilterThroughWindow.java [hunk 3]
```diff
@@ -81,88 +55,31 @@ public Rule build() {
...
-                Preconditions.checkArgument(leftChild instanceof SlotReference
-                        && rightChild instanceof IntegerLikeLiteral);
-
-                long limitVal = ((IntegerLikeLiteral) rightChild).getLongValue();
-                // Adjust the value for 'limitVal' based on the comparison operators.
-                if (conjunct instanceof LessThan) {
-                    limitVal--;
+            Set<Expression> commonPartitionKeys = window.getCommonPartitionKeyFromWindowExpressions();
+            Set<Expression> bottomConjuncts = Sets.newHashSet();
+            Set<Expression> upperConjuncts = Sets.newHashSet();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushdownFilterThroughWindow.java:60]: use ImmutableSet.Builder to construct set

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushdownFilterThroughWindow.java [hunk 4]
```diff
@@ -81,88 +55,31 @@ public Rule build() {
...
+            if (bottomConjuncts.isEmpty()) {
+                return null;
             }
 
-            Optional<Plan> newWindow = window.pushPartitionLimitThroughWindow(partitionLimit, false);
-            if (!newWindow.isPresent()) {
-                return filter;
-            }
-            return filter.withChildren(newWindow.get());
+            LogicalFilter bottomFilter = new LogicalFilter(bottomConjuncts, window.child());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushdownFilterThroughWindow.java:78]: do not use Raw class
```suggestion
            LogicalFilter<Plan> bottomFilter = new LogicalFilter<>(bottomConjuncts, window.child());
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalWindow.java (hunk 2 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalWindow.java:241]: this is typo-error. I will fix it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/RuleSet.java [hunk 5]
```diff
@@ -128,14 +129,16 @@ public class RuleSet {
             new PushdownFilterThroughAggregation(),
             new PushdownFilterThroughRepeat(),
             new PushdownFilterThroughSetOperation(),
-            new PushdownFilterThroughWindow(),
+            new PushdownPartitionTopNThroughWindow(),
```
[the remark this answers is not in the document]
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/RuleSet.java:132]: good name

**Comment by englefly** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushdownFilterThroughWindow.java [hunk 6]
```diff
@@ -81,88 +55,36 @@ public Rule build() {
...
-                }
-                if (limitVal < 0) {
-                    return new LogicalEmptyRelation(ctx.statementContext.getNextRelationId(), filter.getOutput());
+            Set<Expression> commonPartitionKeys = window.getCommonPartitionKeyFromWindowExpressions();
+            Set<Expression> bottomConjuncts = Sets.newHashSet();
+            Set<Expression> upperConjuncts = Sets.newHashSet();
+            for (Expression expr : filter.getConjuncts()) {
+                boolean pushed = false;
+                for (Expression partitionKey : commonPartitionKeys) {
+                    if (partitionKey.getInputSlots().containsAll(expr.getInputSlots())) {
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushdownFilterThroughWindow.java:64]: select * from (select c.*, rank() over(partition by c2+c3 order by c4) rn from c) abc where c2=1;
Seems that c2=1 can't be pushed down. We can just support column currently.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushdownFilterThroughWindow.java
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushdownFilterThroughWindow.java:1]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushdownFilterThroughWindow.java (hunk 6 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushdownFilterThroughWindow.java:64]: thanks

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Review by wm1581066 — APPROVED** [review]
LGTM

**Review by starocean999 — APPROVED** [review]

**Pull request merged**